### PR TITLE
[DI] dont mandate a class on inline services with a factory

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

This check is too strict, useless and boring for inline services :)

### CODE REVIEW

**Comment by stof** [comment]
should this be covered by a test to prevent regressions ? I think the answer should be yes.

File: src/Symfony/Component/DependencyInjection/Compiler/CheckDefinitionValidityPass.php [hunk 1]
```diff
@@ -43,7 +43,7 @@ public function process(ContainerBuilder $container)
             }
 
             // non-synthetic, non-abstract service has class
-            if (!$definition->isAbstract() && !$definition->isSynthetic() && !$definition->getClass()) {
+            if (!$definition->isAbstract() && !$definition->isSynthetic() && !$definition->getClass() && (!$definition->getFactory() || '.' !== substr($id, 0, 1))) {
```
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckDefinitionValidityPass.php]: Any reason to use `substr()` instead of `$id[0]`?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/CheckDefinitionValidityPass.php]: updated to a stricter regexp actually

**Comment by nicolas-grekas** [comment]
Now with tests.

**Review by javiereguiluz — APPROVED** [review]

**Pull request merged**